### PR TITLE
[Test] Add custom Rule to retry failed tests if annotated appropriately.

### LINKED ISSUE #20461: Add retry mechanism for flaky UI tests.
### Expected behavior

Flaky tests fail without any retry logic.

### Actual behavior

When a test should work, but flakiness occurs, retry it N number of times before failing.

### Steps to reproduce the behavior

StatsTests are flaky due to a card not appearing sometimes. Reason unknown.

##### Tested on [device], Android [version], JPAndroid / WPAndroid [version]

Pixel 7 API 34 Emulator

### PR DESCRIPTION
**Pull request by notandyvee**
Fixes #20461 

We had a long running flaky tests. The reasoning for the fail is unclear and it was turned off (https://github.com/wordpress-mobile/WordPress-Android/pull/20493). In general retrying the test appeared to make it pass. Instead of ignoring the test we should attempt to auto retry running the failing tests since odds are it would pass a second time.

-----

## To Test:

- [ ] If builds pass we should be good.

-----

## Regression Notes

1. Potential unintended areas of impact

    N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    Ran other tests.

3. What automated tests I added (or what prevented me from doing so)

    Added TestRule to fix ignored test.

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by irfano** [comment]
I'm thinking about whether adding a retry mechanism to UI tests is the right approach. 🤔  This might prevent us from detecting real issues in production. 
Do you know the root cause of flakiness in the `e2eAllDayStatsLoad()` test, @notandyvee?

**Comment by notandyvee** [comment]
> I'm thinking about whether adding a retry mechanism to UI tests is the right approach

UI tests can be flaky. Some more than others. I can't reproduce the issue. Seems to only happen in UI tests tbh. But, please note this does not mask the issue. If it's a legitimate issue the test will fail and the failure will show. While running all of the UI tests I noticed other failures completely unrelated. The nature of UI tests. If we think it's not worth it, I'm fine closing.

edit: 
> Do you know the root cause of flakiness in the e2eAllDayStatsLoad() test

I can't reproduce. But see this internal convo for some more context: p1710340006238599/1709544863.226669-slack-C5ALGJYEP

**Comment by notandyvee** [comment]
Thoughts @pachlava ?

**Comment by pachlava** [comment]
> I'm thinking about whether adding a retry mechanism to UI tests is the right approach.

@irfano 👋 On iOS, the default retry for UI tests is 2 across several apps, so this approach is in use. UI tests indeed might be flaky, especially on CI (Firebase), where the performance is much lower, and in some cases our attempts to reproduce the fails locally end with nothing.

>Thoughts @pachlava ?

Having a retry mechanism in place is a good thing, IMO - thank you for this, @notandyvee 🙇 

However, with this particular test, it's not about its flakiness. It's about the app not loading certain part of the stats screen, and the test rightfully failing, because it fails to find these parts. This happens only in JP, and I could never reproduce this on a real device. One could think that this is about UI tests environment (different config / using WireMock) causing this, but then this would also occur in WP. This never occurred in WP.

It'll be hard to find a discussion, because it's at least one year old, but the issue about Emulator not always loading the Days stats card in JP is a known issue, but of such a low severity (it affects 0 users to my knowledge), that it was not handled. Taking this all into account, I have no objections to at least trying this test with a rerun.

File: WordPress/src/androidTest/java/org/wordpress/android/rules/Retry.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.wordpress.android.rules
+
+/**
+ * Annotation used to denote you want to retry a UI test function.
+ *
+ * @property value the number of times you want to retry the function, with a default of 1.
+ */
+@Retention(AnnotationRetention.RUNTIME)
+@Target(AnnotationTarget.FUNCTION)
+annotation class Retry(val value: Int = 1)
```
Comment (pachlava) [inline WordPress/src/androidTest/java/org/wordpress/android/rules/Retry.kt]: WDYT of renaming `value` to `reruns` or `numberOfTimes`? This way, it'll be absolutely clear what this parameter does. Now, even though it's more or less intuitive, `value` might give an impression of being used as a boolean analogy only (1 - rerun / 0 - no reruns).

**Review by pachlava — APPROVED** [review]
@notandyvee 👋 

Kudos for implementing this. Regardless of its necessity for a specific stats test (even though I think - https://github.com/wordpress-mobile/WordPress-Android/pull/20517#issuecomment-2020070036 - it's worth trying), having this option is definitely good for future use.

I did some local testing, which included:
- Running the test against WP and making sure it's attempted for execution and then skipped only once, not twice.
- Running against JP, with default `value = 1`, and waiting until the card with stats does not get loaded on the first run, seeing test re-executed and passing on the 2nd try, when the card was loaded. (Which is the real-life use case and reason for the rerun parameter).
- Running against JP with injected failure, and removing `value` (which defaults to 1 rerun), also using `0` and `2` for value, and seeing the number of reruns being always correct.

So, from my side, it's good to go 👍. I left one minor naming note. I didn't do an in-depth code review of `RetryTestRule`, though.

File: WordPress/src/androidTest/java/org/wordpress/android/rules/Retry.kt (hunk 1 above)
Comment (notandyvee) [reply WordPress/src/androidTest/java/org/wordpress/android/rules/Retry.kt]: I think this is a fair ask. I'll update the naming plus default it to retry 2 times to mirror iOS. And then we can see how this works. I'll push an update soon.

**Comment by irfano** [comment]
>Seems to only happen in UI tests tbh.

> UI tests indeed might be flaky, especially on CI (Firebase), where the performance is much lower, and in some cases our attempts to reproduce the fails locally end with nothing.

If the reason for flakiness is the testing environment or emulator performance, then rerun makes sense.

Also, Buildkite mentioned [rerun failing tests](https://buildkite.com/blog/understanding-flaky-tests) in its document but it recommends it after the `Identify the cause of failure and fix it` step. That's why I questioned adding rerun ability.

> One could think that this is about UI tests environment (different config / using WireMock) causing this, but then this would also occur in WP. This never occurred in WP.

@pachlava, this is expected because stats is disabled on WP, so `StatsGranularTabsTest` always passes for WP.

> It'll be hard to find a discussion, because it's at least one year old, but the issue about Emulator not always loading the Days stats card in JP is a known issue, but of such a low severity (it affects 0 users to my knowledge), that it was not handled.

I could reproduce https://github.com/wordpress-mobile/WordPress-Android/issues/18065#issuecomment-2021562993 on a real device with a real site. So, I think this is a real issue and should not be masked with a rerun. If you agree, we can close this PR. WDYT, @pachlava, @notandyvee?

**Comment by pachlava** [comment]
> @pachlava, this is expected because stats is disabled on WP, so StatsGranularTabsTest always passes for WP.

Sorry for not providing more background @irfano. Because we mock the API, the stats are actually available in WP even now, when used with WireMock. And this test never fails in WP because of not-loading card, as well it never failed when stats were still 'officially' present in WP. That's the whole reason why this test was enabled for WP and disabled for JP - we knew the loading issue occurs in JP only, and even though we knew that running this stats test in WP is fully artificial, we decided that it's better to have the test running there this way, than not running it at all. It was so until recently, when @notandyvee  [toggled](https://github.com/wordpress-mobile/WordPress-Android/commit/375a15ff2b30b73e11afe1245fbdf5bf09cf26c7#diff-62f7d46f08918e48d2cfff2a8ea3eb0fc9f229e1e633547e91b95f5773e249a4R27) the execution rule (at that time the test was disabled for both, but because of a different reason I believe). Here's a video I made today, with the latest WP from `trunk`, running against a mocked API:

https://github.com/wordpress-mobile/WordPress-Android/assets/73365754/1547a602-8a5c-4c28-8844-91531e489d68

This is the expected behaviour I believe, and not a bug, because this happens under a mocked API only.

In any case, since you were able to reproduce the issue on a real device (which is very nice ⭐), it will hopefully get fixed

**Comment by notandyvee** [comment]
> I could reproduce https://github.com/wordpress-mobile/WordPress-Android/issues/18065#issuecomment-2021562993 on a real device with a real site. So, I think this is a real issue and should not be masked with a rerun. If you agree, we can close this PR. WDYT, @pachlava, @notandyvee?

First, I think we should merge this regardless @irfano. I can remove the `@Retry` on the test itself if you want. Super useful to have this retry mechanism. Just my opinion.

But I also started to investigate this problem. And figured out how to make it happen reliably. I am trying to find the root cause. In the meantime I will add a needs revision flag while i finish the investigation.

**Comment by irfano** [comment]
> First, I think we should merge this regardless @irfano. I can remove the @Retry on the test itself if you want. Super useful to have this retry mechanism. Just my opinion.

Good decision, I agree! 👍🏻 We can merge this after removing `@Retry` from `e2eAllDayStatsLoad()`.

**Comment by notandyvee** [comment]
I went ahead and removed the `@Retry` for now and re-added the ignore since I am investigating the root cause. The linked issue is done though.

**Pull request merged**